logger.go

```diff
@@ -5,155 +5,260 @@ import (
 	"os"
 	"runtime"
 	"strings"
+	"sync"
+	"testing"
 	"time"
 )
 
-type logWrappedT struct {
-	T
+type loggerT[ET T] struct {
+	runTHelper    // Embeds T and provides Fail/Parallel
+	orig       ET // Keep reference to original for Run method
+	logger     func(string)
+	skipFrames int // Additional skip frames for nested wrappers
+}
+
+// replaceLoggerT directly implements T interface to avoid double loggerT wrapping
+type replaceLoggerT[ET T] struct {
+	runTHelper
+	orig   ET
 	logger func(string)
 }
 
-// ReplaceLogger creates a T that is wrapped so that the logger is
-// overridden with the provided function.
-func ReplaceLogger(t T, logger func(string)) T {
-	return logWrappedT{
-		T:      t,
-		logger: logger,
+// ReplaceLogger creates a logger wrapper that overrides the logging function. When layered
+// on top of BufferedLogger, it assumes that only one extra stack frame is added. If that's
+// not the case, cast and adjust:
+//
+//	if asf, ok := t.(interface{ AdjustSkipFrames(int) }); ok {
+//		asf.AdjustSkipFrames(2)
+//	}
+//
+// This adjustment should be done before using the the returned T
+func ReplaceLogger[ET T](t ET, logger func(string)) T {
+	// If the underlying logger supports AdjustSkipFrames, adjust it to account for
+	// the extra call frames: replaceLoggerT.Log -> custom logger function -> underlying logger call
+	if adjuster, ok := any(t).(interface{ AdjustSkipFrames(int) }); ok {
+		adjuster.AdjustSkipFrames(2) // +2 for replaceLoggerT.Log and the custom logger function
+	}
+
+	wrapped := &replaceLoggerT[ET]{
+		runTHelper: runTHelper{T: t},
+		orig:       t,
+		logger:     logger,
 	}
+	return any(wrapped).(T)
 }
 
-func (t logWrappedT) Log(args ...interface{}) {
+func (t replaceLoggerT[ET]) Log(args ...interface{}) {
 	line := fmt.Sprintln(args...)
-	t.logger(line[0 : len(line)-1])
+	message := line[0 : len(line)-1]
+	t.logger(message)
 }
 
-func (t logWrappedT) Logf(format string, args ...interface{}) {
-	t.logger(fmt.Sprintf(format, args...))
+func (t replaceLoggerT[ET]) Logf(format string, args ...interface{}) {
+	message := fmt.Sprintf(format, args...)
+	t.logger(message)
 }
 
-// AdjustSkipFrames adjusts skip frames on the underlying T if it supports it
-func (t logWrappedT) AdjustSkipFrames(skip int) {
-	if adjuster, ok := t.T.(interface{ AdjustSkipFrames(int) }); ok {
-		adjuster.AdjustSkipFrames(skip)
+// Run implements the new RunT interface
+func (t replaceLoggerT[ET]) Run(name string, f func(*testing.T)) bool {
+	if runnable, ok := any(t.orig).(interface {
+		Run(string, func(*testing.T)) bool
+	}); ok {
+		return runnable.Run(name, f)
```

Comment:
This technique of checking at runtime whether an interface is supported make me wonder if there could be just one `loggerT` type, and do the same thing in the `Log` and `Logf` methods.